aptos-move/mvhashmap/src/versioned_data.rs

```diff
@@ -115,14 +133,110 @@ impl<V: TransactionWrite> Default for VersionedValue<V> {
     }
 }
 
-impl<V: TransactionWrite> VersionedValue<V> {
-    fn read(&self, txn_idx: TxnIndex) -> anyhow::Result<MVDataOutput<V>, MVDataError> {
+// TODO(BlockSTMv2): remove TransactionWrite trait requirement from V, even if
+// AggregatorV1 code is not removed by definining a more specialized trait that can
+// runtime assert in other variants. Add other variants to stored cells that allow
+// storing group metadata and group size together at the group key (currently metadata
+// is stored in a fake write entry, and group size is stored in MVGroupData map).
+impl<V: TransactionWrite + PartialEq> VersionedValue<V> {
+    // Extracts read dependencies from the data-structure that are affected by a write
+    // of 'data' at 'txn_idx'. Some of these dependencies that remain valid can be
+    // relocated by a caller to a different (new) entry in the data-structure. The
+    // boolean flag indicates whether the dependency is still valid after the write.
+    fn get_affected_read_dependencies<const ONLY_COMPARE_METADATA: bool>(
+        &self,
+        txn_idx: TxnIndex,
+        data: &Arc<V>,
+        maybe_layout: &Option<Arc<MoveTypeLayout>>,
+    ) -> (BTreeSet<(TxnIndex, Incarnation)>, bool) {
+        let mut affected_deps = BTreeSet::new();
+        let mut validation_result = true;
+
+        // Look at entries at or below txn_idx
+        if let Some((_, entry)) = self
+            .versioned_map
+            .range(..=ShiftedTxnIndex::new(txn_idx))
+            .next_back()
+        {
+            if let EntryCell::ResourceWrite(_, value, dependencies) = &entry.value {
+                // Take dependencies above txn_idx
+                let mut deps = dependencies.lock();
+                let new_deps = deps.split_off(&(txn_idx + 1, 0));
+                if !new_deps.is_empty() {
+                    // Only validate if we have dependencies to check.
+                    // Validate by comparing both value and layout, or metadata if only_compare_metadata is true
+                    validation_result = if let ValueWithLayout::Exchanged(v, l) = value {
```

Comment:
nit: right now this is because we have `RawFromStorage` which should not be validated. Do you think it is safer to have an explicit match?